src/MixedMatrix.cpp

```diff
@@ -15,221 +15,311 @@
 {
+    swap(lhs.edge_true_edge_, rhs.edge_true_edge_);
 
+    swap(lhs.M_local_, rhs.M_local_);
+    swap(lhs.D_local_, rhs.D_local_);
+    swap(lhs.W_local_, rhs.W_local_);
+
+    swap(lhs.M_global_, rhs.M_global_);
+    swap(lhs.D_global_, rhs.D_global_);
+    swap(lhs.W_global_, rhs.W_global_);
+
+    std::swap(lhs.offsets_, rhs.offsets_);
+    std::swap(lhs.true_offsets_, rhs.true_offsets_);
+
+    swap(lhs.M_elem_, rhs.M_elem_);
+    swap(lhs.elem_dof_, rhs.elem_dof_);
+
+    swap(lhs.agg_vertexdof_, rhs.agg_vertexdof_);
+    std::swap(lhs.face_facedof_, rhs.face_facedof_);
 }
-MixedMatrix::MixedMatrix(std::unique_ptr<mfem::SparseMatrix> M,
-                         std::unique_ptr<mfem::SparseMatrix> D,
-                         std::unique_ptr<mfem::SparseMatrix> W,
-                         const mfem::HypreParMatrix& edge_d_td)
-    :
-    M_(std::move(M)),
-    D_(std::move(D)),
-    W_(std::move(W)),
-    edge_d_td_(&edge_d_td),
-    edge_td_d_(edge_d_td_->Transpose())
+
+int MixedMatrix::Rows() const
 {
-    GenerateRowStarts();
+    return D_local_.Rows() + D_local_.Cols();
 }
 
-/// @todo better documentation of the 1/-1 issue, make it optional?
-void MixedMatrix::Init(const mfem::SparseMatrix& vertex_edge,
-                       const mfem::Vector& weight,
-                       const mfem::SparseMatrix& w_block)
+int MixedMatrix::Cols() const
 {
+    return D_local_.Rows() + D_local_.Cols();
+}
 
+int MixedMatrix::GlobalRows() const
+{
+    return D_global_.GlobalRows() + D_global_.GlobalCols();
+}
 
-    const mfem::HypreParMatrix& edge_d_td(*edge_d_td_);
-    const int nvertices = vertex_edge.Height();
-    const int nedges = vertex_edge.Width();
+int MixedMatrix::GlobalCols() const
+{
+    return D_global_.GlobalRows() + D_global_.GlobalCols();
+}
 
-    int* M_fine_i = new int [nedges + 1];
-    int* M_fine_j = new int [nedges];
-    double* M_fine_data = new double [nedges];
-    std::iota(M_fine_i, M_fine_i + nedges + 1, 0);
-    std::iota(M_fine_j, M_fine_j + nedges, 0);
-    std::copy_n(weight.GetData(), nedges, M_fine_data);
+int MixedMatrix::NNZ() const
+{
+    return M_local_.nnz() + (2 * D_local_.nnz())
+           + W_local_.nnz();
+}
 
-    for (int i = 0; i < nedges; i++)
-    {
-        assert(M_fine_data[i] != 0.0);
-        M_fine_data[i] = 1.0 / std::fabs(M_fine_data[i]);
-    }
+int MixedMatrix::GlobalNNZ() const
+{
+    return M_global_.nnz() + (2 * D_global_.nnz())
+           + W_global_.nnz();
+}
 
-    M_ = make_unique<mfem::SparseMatrix>(M_fine_i, M_fine_j, M_fine_data,
-                                         nedges, nedges);
+bool MixedMatrix::CheckW() const
+{
+    int local_size = W_global_.Rows();
+    int global_size;
+
+    MPI_Allreduce(&local_size, &global_size, 1, MPI_INT, MPI_MAX, D_global_.GetComm());
+
+    const double zero_tol = 1e-6;
+
+    return global_size > 0 && W_global_.MaxNorm() > zero_tol;
+}
+
+
+ParMatrix MixedMatrix::ToPrimal() const
```

Comment:
Currently this is function is only called in `generalgraph.cpp` (correct me if I missed something). I am thinking to use this function also in `MinresBlockSolver` and `SPDSolver` because in theory they need the same operation, some adjustment is needed though.